staging/src/k8s.io/api/admissionregistration/v1beta1/types.go

```diff
@@ -576,31 +592,587 @@ type ServiceReference struct {
+	// the failureMode applies.
+	//
+	// `PerNamespace` and `ClusterWide` are members of a union; if one of the
+	// fields is set, the other must be unset.
+	//
+	// +optional
+	PerNamespace *NamespaceParamRef `json:"perNamespace,omitempty" protobuf:"bytes,3,opt,name=perNamespace"`
 }
+
+// BindingParamKind is the kind of ValidatingAdmissionPolicyBinding.
+// +enum
+type BindingParamKind string
+
+// Supported binding param kinds.
+const (
+	BindingParamKindClusterWide  BindingParamKind = "ClusterWide"
+	BindingParamKindPerNamespace BindingParamKind = "PerNamespace"
+)
+
+// ClusterWideParamRef references a parameter resource to be used as input
+// to all resource requests.
+// +structType=atomic
+type ClusterWideParamRef struct {
+	// Name of the resource being referenced.
+	Name string `json:"name,omitempty" protobuf:"bytes,1,rep,name=name"`
+	// Namespace of the referenced resource.
+	// Should be empty for the cluster-scoped resources
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,2,rep,name=namespace"`
+}
+
+// NamespaceParamRef references parameters resource scoped to the same
+// namespace as the resource being evaluated by a policy. The parameters
+// must all be of the kind specified in the policy's `paramKind` field.
+//
+// If multiple objects are found to match, the policy will be evaluated
+// against ALL of them and the result will be ANDed together.
+//
+// If no objects match the selector, then the `failAction` applies.
+//
+// If all fields (other than `failAction`) are empty, then the policy will be
+// evaluated against ALL resources of the policy's `paramKind` in the request's
+// namespace.
+//
+// +structType=atomic
+type NamespaceParamRef struct {
+	// Name is the name of the namespace-scoped param to evaluate against.
+	// When evaluating the admission of a resource request, a param with the
+	// given name will be found in the same namespace as the resource and used
+	// as the `param` input to the policy's rules.
+	//
+	// +optional
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+
+	// Selector can be used to match multiple param objects in the namespace
+	// based on metadata.
+	//
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty" protobuf:"bytes,2,opt,name=selector"`
+
+	// FailAction controls the behavior of the binding when the resource
+	// exists, and name or selector is valid, but there are no parameters
+	// matched by the binding. If the value is set to `Allow`, then no
+	// matched parameters will be treated as successful validation by the binding.
+	// If set to `Deny`, then no matched parameters will be subject to the
+	// `failurePolicy` of the policy.
+	//
+	// Allowed values are `Allow` or `Deny`
+	// Default to `Allow`
+	// +optional
+	FailAction *ParamFailureActionType `json:"failAction,omitempty" protobuf:"bytes,3,opt,name=failAction,casttype=ParamFailureActionType"`
```

Comment:
```suggestion
	ParameterNotFoundAction *ParameterNotFoundActionType `json:"parameterNotFoundAction,omitempty" protobuf:"bytes,3,opt,name=parameterNotFoundAction,casttype=ParameterNotFoundActionType"`
```

and other versions.

The selection itself was valid, but there were no matches.

Reply:
renamed as suggested